Stack/Opc.Ua.Core/Security/Certificates/CertificateValidator.cs

```diff
@@ -691,8 +692,60 @@ private void SaveCertificates(X509Certificate2Collection certificateChain)
                         ICertificateStore store = m_rejectedCertificateStore.OpenStore();
                         try
                         {
+                            var certs = store.Enumerate().GetAwaiter().GetResult();
+                            int existingCertsCount = certs.Count;
+                            var certsThumbprints = new HashSet<string>(certs.Cast<X509Certificate2>().Select(c => c.Thumbprint));
+
+                            // Do not add existing certificates from the chain
+                            var newChainCerts = certificateChain.Cast<X509Certificate2>().Where(c => !certsThumbprints.Contains(c.Thumbprint)).ToArray();
+                            int newChainCertsCount = newChainCerts.Length;
+                            // Nothing to add
+                            if (newChainCertsCount == 0)
+                            {
+                                Utils.LogTrace("The entire rejected certificate chain is allready present in {0}", m_rejectedCertificateStore);
+                                return;
+                            }
+
+                            X509Certificate2Collection certsToRemove = null;
+                            X509Certificate2Collection certsToAdd = new X509Certificate2Collection();
+
+                            // If newChainCerts is multiple of kMaxRejectedCerts + r
+                            // all the existing certificates need to be cleared and only last kMaxRejectedCerts will be added
+                            if (newChainCertsCount >= kMaxRejectedCerts)
+                            {
+                                certsToRemove = certs;
+
+                                int startIdx = newChainCertsCount - kMaxRejectedCerts;
+                                for (int i = startIdx; i < newChainCertsCount; i++)
+                                {
+                                    certsToAdd.Add(newChainCerts[i]);
+                                }
+                            }
+                            else
+                            {
+                                certs.AddRange(newChainCerts);
+
+                                int startIdx = certs.Count - kMaxRejectedCerts;
+
+                                if (startIdx > 0)
+                                {
+                                    if (startIdx <= existingCertsCount)
+                                    {
+                                        certsToRemove = new X509Certificate2Collection();
+                                        for (int i = 0; i < startIdx; i++)
+                                        {
+                                            certsToRemove.Add(certs[i]);
```

Comment:
don´t we need to store the certs to delete oldest in the rejected store first, we cant rely on the cert store to return them in this order, however i am not shure how to achvieve this without file creation date?

Reply:
correct